src/Microsoft.ML.FastTree/TreeEnsemble/RegressionTree.cs

```diff
@@ -14,25 +14,73 @@
 using Microsoft.ML.Model;
 using Microsoft.ML.Model.Pfa;
 using Newtonsoft.Json.Linq;
-using Float = System.Single;
 
 namespace Microsoft.ML.Trainers.FastTree.Internal
 {
-    public class RegressionTree
+    public class RegressionTreeView
     {
-        private double _maxOutput;
+        /// <summary>
+        /// <see cref="RegressionTreeView"/> is an immutable wrapper over <see cref="_tree"/> for exposing some tree's attribute to users.
+        /// </summary>
+        private readonly RegressionTree _tree;
+
+        // Immutable accessors to the underlying tree structure. They help users to inspect tree learned using, for example,
+        // gradient-boosting decision tree and random forest.
+        public ReadOnlySpan<int> LteChild => _tree.LteChild;
+        public ReadOnlySpan<int> GtChild => _tree.GtChild;
+        public ReadOnlySpan<int> NumericalSplitFeatureIndexes => _tree.SplitFeatures;
+        public ReadOnlySpan<float> NumericalSplitThresholds => _tree.RawThresholds;
+        public ReadOnlySpan<bool> CategoricalSplitFlags => _tree.CategoricalSplit;
+        public IReadOnlyList<IReadOnlyList<int>> CategoricalSplitFeatures => _tree.CategoricalSplitFeatures;
+        public IReadOnlyList<IReadOnlyList<int>> CategoricalSplitFeatureRanges => _tree.CategoricalSplitFeatureRanges;
+        public ReadOnlySpan<double> LeafValues => _tree.LeafValues;
+
+        public IReadOnlyList<IReadOnlyList<double>> LeafSamples { get; }
+        public IReadOnlyList<IReadOnlyList<double>> LeafSampleWeights { get; }
 
-        // Weight of this tree in the ensemble
+        /// <summary>
+        /// Number of leaves in the tree. Note that <see cref="NumLeaves"/> does not take non-leaf nodes into account.
+        /// </summary>
+        public int NumLeaves => _tree.NumLeaves;
 
-        // for each non-leaf, we keep the following data
-        public Float[] DefaultValueForMissing;
+        /// <summary>
+        /// Number of nodes in the tree. This doesn't include any leaves. For example, a tree with node0->node1, node0->leaf3,
+        /// node1->leaf1, node1->leaf2, <see cref="NumNodes"/> and <see cref="NumLeaves"/> should be 2 and 3, respectively.
+        /// </summary>
+        // A visualization of the example mentioned in this doc string.
+        //         node0
+        //         /  \
+        //     node1 leaf3
+        //     /  \
+        // leaf1 leaf2
+        public int NumNodes => NumLeaves - 1;
+
+        internal RegressionTreeView(RegressionTree tree)
+        {
+            _tree = tree;
+            LeafSamples = null;
+            LeafSampleWeights = null;
+            if (tree is QuantileRegressionTree)
+            {
+                LeafSamples = ((QuantileRegressionTree)tree).ExtractLeafSamples();
+                LeafSampleWeights = ((QuantileRegressionTree)tree).ExtractLeafSampleWeights();
+            }
+        }
+    }
+
+    public class RegressionTree
```

Comment:
>public [](start = 4, length = 6)

Probably internal.

Reply:
It will be also best-friend because of being used in tests.

---
In reply to: [251182509](https://github.com/dotnet/machinelearning/pull/2243#discussion_r251182509) [](ancestors = 251182509)